src/transformer.jl

```diff
@@ -87,91 +84,147 @@ function RunState(config::Config)
+This modifies the RunState of the transformer.
 
-llama2.c correspondence: read_config (l. 147)
+llama2.c correspondence: forward (l. 231)
 """
+function forward!(transformer::Transformer, token::Int, pos::Int)::Array{Float32}
+    # a few convenience variables
+    (; dim, n_heads, n_kv_heads, n_layers, seq_len) = transformer.config
+    1 <= pos <= seq_len || throw(ArgumentError("1 <= pos <= seq_len is required"))
+    s = transformer.state
+    w = transformer.weights
+    kv_dim = (dim * n_kv_heads) ÷ n_heads
+    kv_mul = n_heads ÷ n_kv_heads # integer multiplier of the kv sharing in multiquery
+    head_size = dim ÷ n_heads
+
+    # copy the token embedding into x
+    s.x = w.token_embedding_table[:, token] # (dim,)
+
+    # forward all the layers
+    for l in 1:n_layers
+        # attention rmsnorm
+        s.xb = rmsnorm(s.x, w.rms_att_weight[:, l]) # (n_heads * head_size,)
+
+        # qkv matmuls for this position
+        s.q = w.wq[:, :, l]' * s.xb # (n_heads * head_size,) = (dim, n_heads * head_size) * (n_heads * head_size,)
+        s.key_cache[:, pos, l] = w.wk[:, :, l]' * s.xb # (kv_dim,) = (kv_dim, dim) * (n_heads * head_size,)
+        s.value_cache[:, pos, l] = w.wv[:, :, l]' * s.xb # (kv_dim,) = (kv_dim, dim) * (n_heads * head_size,)
+
+        # RoPE relative positional encoding: complex-valued rotate q and k in each head
+        for i in 1:2:dim
+            head_dim = (i - 1) % head_size
+            freq = 1.0f0 / (10000.0f0^(head_dim / head_size))
+            val = (pos - 1) * freq
+            fcr = cos(val)
+            fci = sin(val)
+
+            # rotate query
+            v0 = s.q[i]
+            v1 = s.q[i + 1]
+            s.q[i] = v0 * fcr - v1 * fci
+            s.q[i + 1] = v0 * fci + v1 * fcr
+
+            # rotate key only if i <= kv_dim
+            if i <= kv_dim
+                v0 = s.key_cache[i, pos, l]
+                v1 = s.key_cache[i + 1, pos, l]
+                s.key_cache[i, pos, l] = v0 * fcr - v1 * fci
+                s.key_cache[i + 1, pos, l] = v0 * fci + v1 * fcr
+            end
+        end
+
+        # iterate over all heads
+        for h in 0:(n_heads - 1)
+            # get the query vector for this head
+            h_off = h * head_size
+            q = s.q[(h_off + 1):(h_off + head_size)] # (head_size,)
+            # iterate over all timesteps, including the current one
+            kv_ind = (h ÷ kv_mul) * head_size
+            for t in 1:pos
+                k = s.key_cache[(kv_ind + 1):(kv_ind + head_size), t, l] # (head_size,)
+                score = (q ⋅ k) / sqrt(Float32(head_size)) # scalar
+                s.att[h + 1, t] = score
+            end
+            # softmax the scores to get attention weights, from 1..pos inclusively
+            s.att[h + 1, 1:pos] = softmax(s.att[h + 1, 1:pos]) # (pos,)
+            # weighted sum of the values, store back into xb
+            s.xb[(h_off + 1):(h_off + head_size)] .= 0 # (head_size,)
+            for t in 1:pos
```

Comment:
Also here, shouldnt it be: for t in 1:(pos + 1)
Because in llama2.c it says: t <= pos

Reply:
No, as pos is 1-based